2022/Q4/artifacts/PoS-pseudocode/PoS-model.md

```diff
@@ -17,6 +17,11 @@ type JailRecord struct {
   epoch Epoch
 }
 
+type UnbondRecord struct {
+  amount uint
+  start Epoch
+}
+
 type Validator struct {
```

Comment:
`total_unbonds` should really be `total_unbonded`, correct?

Reply:
Do you mean the field in the Validator's struct? If I am not wrong, we use `total_unbonds` throughout the spec and `total_unbonded` at the `end-of-epoch` function when folding `total_unbonds`. I agree we could find better names. Was that your concern, or I misunderstood it?

Reply:
Ok thanks for clarifying that these are intended to be different.

I'll elaborate a bit. First, yes, the field `total_unbonds` inside the `Validator` struct should be a `map<Epoch, UnbondRecord>` judging by the `unbond` code. It also seems that the `Validator` struct should explicitly have a `total_unbonded` field that is a `map<Epoch, uint>`, right? I think I was confused by the lack of a declaration of `total_unbonded` anywhere outside the `end_of_epoch()` code.

I'm still somewhat confused by what the purpose of `Validator::total_unbonds` is. From the pseudocode, it seems only to be used in `func unbond`, where we set values inside of it, but I don't see any other place where its data is read and then used for something. It is not in `end_of_epoch`, but perhaps there is another bug. From the following code from `end_of_epoch`:

```
      var total_unbonded = 0
      //find the total unbonded from the slash epoch up to the current epoch first
      //a..b notation determines an integer range: all integers between a and b inclusive
      forall (epoch in slash.epoch+1..cur_epoch) do
        forall (unbond in validators[validator_address].total_unbonded[epoch] s.t. unbond.start <= slash.epoch)
          total_unbonded += unbond.amount
```
it looks like perhaps `forall (unbond in validators[validator_address].total_unbonded[epoch]` should be `forall (unbond in validators[validator_address].total_unbonds[epoch]`. Does this seem to be the intention?

Reply:
actually I think I'm realizing that `total_unbonds` should be a `map<Epoch, Set<UnbondRecord>>`, where `Set<UnbondRecord>` could also be a `Vec` or similar